pkg/sql/sem/tree/eval.go

```diff
@@ -4421,21 +4437,60 @@ func replaceCustomEscape(s string, escape rune) (string, error) {
 				panic("unexpected: escape character is the last one in replaceCustomEscape.")
 			}
 		} else if s[sIndex] == '\\' {
-			// We encountered a backslash which after QuoteMeta should have been doubled.
-			if sIndex+1 == sLen || s[sIndex+1] != '\\' {
-				// This case should never be reached, and it should
+			// We encountered a backslash, so we need to look ahead to figure out how
+			// to process it.
+			if sIndex+1 == sLen {
+				// This case should never be reached since it should
 				// have been caught in calculateLengthAfterReplacingCustomEscape.
-				panic("unexpected: a backslash is not doubled in replaceCustomEscape.")
+				panic("unexpected: a single backslash encountered in replaceCustomEscape.")
+			} else if s[sIndex+1] == '\\' {
+				// We want to escape '\\' to `\\\\` for correct processing later by unescapePattern. See (3).
+				// Since we've added four characters to ret, we advance retIndex by 4.
+				// Since we've already processed two characters in s, we advance sIndex by 2.
+				ret[retIndex] = '\\'
+				ret[retIndex+1] = '\\'
+				ret[retIndex+2] = '\\'
+				ret[retIndex+3] = '\\'
+				retIndex += 4
+				sIndex += 2
+			} else {
+				// A metacharacter other than a backslash is escaped here.
+				if string(s[sIndex+1]) == string(escape) {
```

Comment:
`s` is a string and `escape` is a rune which means a non-ascii escape value would produce a multibyte `string(escape)` and thus never match the left side. Is this a case we care about?